### ISSUE TITLE
Support for Complex Matrices in Proximal Solvers?

### ISSUE THREAD

**Issue by SHAOSIHAN** [deepinv/deepinv#133]
Hi,

Thanks for your interesting work!

Does the current implementation of the unfolded proximal solver support complex number computations inherently? Such as, 

y = Ax

where y, A and x are complex matrices.

**Comment by matthieutrs** [comment]
Hi, thanks for the interest in our work!

The library can work for complex matrices, but at the moment, not all the functionalities are compatible with complex numbers. For example, when computing the spectral norm of a linear operator in the base `LinearPhysics` class [here](https://github.com/deepinv/deepinv/blob/main/deepinv/physics/forward.py#L339), a modification might be needed. 

We are planning to add support for complex numbers in an imminent future however ;)

**How to bypass this at the moment?**

TL;DR: an example is available in this [colab notebook](https://colab.research.google.com/drive/1lsbEFn-vBMNnD61dbN4KY6tmmLmVS1VE?usp=sharing)

**1. Define your custom operator**

Assume that you have both the forward and (Hermitian) adjoint of your complex operator, namely `my_complex_forward_op` and `my_complex_adjoint_op`. You can implement your `Physics` as such:
```python
import deepinv as dinv

# define an operator that converts color images into grayscale ones.
class MyComplexOperator(dinv.physics.LinearPhysics):
    r"""
    A toy example
    """
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
 
    def A(self, x):
        y = my_complex_forward_op(x)
        return y

    def A_adjoint(y):
        x = my_complex_adjoint_op(y)
        return x

physics = MyComplexOperator()

x = torch.randn((1,1,64,64))+1j*torch.randn((1,1,64,64))
y = physics(x)

```

Once this is done, you can easily access gradient of the operator (note that the proximity operator might need some modifications)

```python
from deepinv.optim.data_fidelity import L2

data_fidelity = L2()

u = torch.randn((1,1,64,64))+1j*torch.randn((1,1,64,64))
x = u - gamma*data_fidelity.grad(u, y, physics)  # Perform a gradient step
```

**2. Wrap the denoiser / proximal operator in a class performing the operations on complex numbers**

```python
import torch
import deepinv as dinv

class ComplexDenoiser(torch.nn.Module):
    def __init__(self, denoiser):
        super().__init__()
        self.denoiser = denoiser

    def forward(self, x, sigma):  # input x is complex
        noisy_batch = torch.cat((x.real, x.imag), 0)  # we split complex data in real and imaginary components
        denoised_batch = self.denoiser(noisy_batch, sigma)  # perform the denoising on the real-valued tensor
        denoised = denoised_batch[0:1, ...]+1j*denoised_batch[1:2, ...]  # build the denoised complex data
        return denoised

# Example: use a wavelet prior and apply it to complex data
denoiser = dinv.models.WaveletPrior(level=3, wv='db8', device='cpu', non_linearity='soft')
model = ComplexDenoiser(denoiser)  # wrap the denoiser in the complex class above


complex_im = torch.randn((1,1,64,64))+1j*torch.randn((1,1,64,64))
denoised = model(complex_im, 0.05)  # denoise
```

Let me know if you want more insight on this :)

**Comment by tachella** [comment]
Support for complex linear operators has been added in this [PR](https://github.com/deepinv/deepinv/pull/168).

Thus, you should be able to use deepinv with complex inverse problems, as long as your model can handle complex numbers (which can be done in a simple way by adding a real and imaginary input channels to the denoising/backbone network)
